### PR TITLE
[TOPIC-GPIO] drivers: hci: spi: update to new GPIO API

### PR DESCRIPTION
**Pull request by erwango**
Update bluetooth hcp spi driver to new GPIO API.

Tested on disco_l475_iot1 and 96b_carbon.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

### CODE REVIEW

**Review by jhedberg — APPROVED** [review]

File: boards/arm/96b_carbon/96b_carbon.dts [hunk 1]
```diff
@@ -84,8 +84,8 @@
 	bt-hci@0 {
 		compatible = "zephyr,bt-hci-spi";
 		reg = <0>;
-		reset-gpios = <&gpiob 4 0>;
-		irq-gpios = <&gpiob 1 0>;
+		reset-gpios = <&gpiob 4 GPIO_ACTIVE_HIGH>;
```
Comment (mnkp) [inline boards/arm/96b_carbon/96b_carbon.dts]: Since the driver is using logical levels to drive the reset signal I believe we should have here `GPIO_ACTIVE_LOW` as the reset is activated when the line is low. The driver code would need to be updated to set value `1` to activate the reset. Otherwise - this is an open point, haven't been discussed so far - we could also have the driver use raw / physical signals. In this case it should use `gpio_pin_set_raw` function to set the pin value and we shouldn't have here `GPIO_ACTIVE_HIGH` flag.

**Review by mnkp — CHANGES_REQUESTED** [review]

**Review by joerchan — APPROVED** [review]

File: drivers/bluetooth/hci/spi.c [hunk 2]
```diff
@@ -479,17 +478,18 @@ static int bt_spi_send(struct net_buf *buf)
...
 	gpio_pin_configure(irq_dev, GPIO_IRQ_PIN,
 #if defined(CONFIG_BT_SPI_BLUENRG)
 			   GPIO_PUD_PULL_DOWN |
 #endif
-			   GPIO_DIR_IN | GPIO_INT |
-			   GPIO_INT_EDGE | GPIO_INT_ACTIVE_HIGH);
+			   GPIO_INPUT);
+
+	gpio_pin_interrupt_configure(irq_dev, GPIO_IRQ_PIN,
+				     GPIO_INT_EDGE_RISING);
```
[the remark this answers is not in the document]
Comment (erwango) [reply drivers/bluetooth/hci/spi.c]: Thanks, so I'm moving the `init/add_callback` function before `gpio_pin_interrupt_configure` then.

**Comment by erwango** [comment]
@mnkp, updated, thanks for the review

**Comment by erwango** [comment]
@mnkp, updated

File: drivers/bluetooth/hci/spi.c [hunk 3]
```diff
@@ -480,26 +482,21 @@ static int bt_spi_open(void)
...
 
 	if (gpio_add_callback(irq_dev, &gpio_cb)) {
 		return -EINVAL;
 	}
 
-	if (gpio_pin_enable_callback(irq_dev, GPIO_IRQ_PIN)) {
-		return -EINVAL;
-	}
+	gpio_pin_interrupt_configure(irq_dev, GPIO_IRQ_PIN,
+				     GPIO_INT_EDGE_RISING);
```
Comment (mnkp) [inline drivers/bluetooth/hci/spi.c]: Since we want to support logical levels (honor `GPIO_ACTIVE_LOW`, `GPIO_ACTIVE_HIGH` flags) we should use here `GPIO_INT_EDGE_TO_ACTIVE` flag.

File: drivers/bluetooth/hci/spi.c [hunk 4]
```diff
@@ -201,22 +204,21 @@ static int configure_cs(void)
 		return -EIO;
 	}
 
-	gpio_pin_configure(cs_dev, GPIO_CS_PIN,
-			   GPIO_DIR_OUT | GPIO_PUD_PULL_UP);
-	gpio_pin_write(cs_dev, GPIO_CS_PIN, 1);
+	gpio_pin_configure(cs_dev, GPIO_CS_PIN, GPIO_OUTPUT | GPIO_CS_FLAGS);
+	gpio_pin_set(cs_dev, GPIO_CS_PIN, 1);
```
Comment (mnkp) [inline drivers/bluetooth/hci/spi.c]: This PR may need to be rebased but latest GPIO API supports `GPIO_OUTPUT_ACTIVE` flag. The two calls can be replaced with
```
	gpio_pin_configure(cs_dev, GPIO_CS_PIN, GPIO_OUTPUT_ACTIVE | GPIO_CS_FLAGS);
```

File: drivers/bluetooth/hci/spi.c [hunk 5]
```diff
@@ -480,26 +482,21 @@ static int bt_spi_open(void)
 {
 	/* Configure RST pin and hold BLE in Reset */
 	gpio_pin_configure(rst_dev, GPIO_RESET_PIN,
-			   GPIO_DIR_OUT | GPIO_PUD_PULL_UP);
-	gpio_pin_write(rst_dev, GPIO_RESET_PIN, 0);
+			   GPIO_OUTPUT | GPIO_RESET_FLAGS);
+	gpio_pin_set(rst_dev, GPIO_RESET_PIN, 1);
```
Comment (mnkp) [inline drivers/bluetooth/hci/spi.c]: `gpio_pin_set` can be removed if `GPIO_OUTPUT_ACTIVE` flag is used.

**Review by mnkp — COMMENTED** [review]
LGTM with minor comments.

**Comment by erwango** [comment]
@mnkp, updated

File: boards/arm/disco_l475_iot1/disco_l475_iot1.dts [hunk 6]
```diff
@@ -118,13 +118,13 @@
 &spi3 {
 	status = "okay";
 
-	cs-gpios = <&gpiod 13 0>, <&gpioe 0 0>;
+	cs-gpios = <&gpiod 13 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP)>, <&gpioe 0 0>;
```
Comment (mnkp) [inline boards/arm/disco_l475_iot1/disco_l475_iot1.dts]: I overlooked it earlier, do we need to add flags also to `gpioe`? Also, this change doesn't seem to be related to hci-spi driver. Maybe we should do it in another PR?

Comment (erwango) [reply boards/arm/disco_l475_iot1/disco_l475_iot1.dts]: `gpioe` is dealt with in https://github.com/zephyrproject-rtos/zephyr/pull/21066.

I see it related in the fact that it deals with the same global functionality which is hci BLE  over SPI. So I have been able to test and validate porting to new API in current change. I don't see much value in splitting it.

**Review by mnkp — APPROVED** [review]
